source/name_mapper.cpp

```diff
@@ -0,0 +1,250 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and/or associated documentation files (the
+// "Materials"), to deal in the Materials without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Materials, and to
+// permit persons to whom the Materials are furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Materials.
+//
+// MODIFICATIONS TO THIS FILE MAY MEAN IT NO LONGER ACCURATELY REFLECTS
+// KHRONOS STANDARDS. THE UNMODIFIED, NORMATIVE VERSIONS OF KHRONOS
+// SPECIFICATIONS AND HEADER INFORMATION ARE LOCATED AT
+//    https://www.khronos.org/registry/
+//
+// THE MATERIALS ARE PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+// IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+// TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+// MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+#include "name_mapper.h"
+
+#include <algorithm>
+#include <iterator>
+#include <sstream>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "spirv-tools/libspirv.h"
+#include "spirv/1.1/spirv.h"
+
+namespace {
+
+// Converts a uint32_t to its string decimal representation.
+std::string to_string(uint32_t id) {
+  // Use stringstream, since some versions of Android compilers lack
+  // std::to_string.
+  std::stringstream os;
+  os << id;
+  return os.str();
+}
+
+}  // anonymous namespace
+
+namespace libspirv {
+
+NameMapper GetTrivialNameMapper() { return to_string; }
+
+FriendlyNameMapper::FriendlyNameMapper(const spv_const_context context,
+                                       const uint32_t* code,
+                                       const size_t wordCount)
+    : grammar_(libspirv::AssemblyGrammar(context)) {
+  spv_diagnostic diag = nullptr;
+  // We don't care if the parse fails.
+  spvBinaryParse(context, this, code, wordCount, nullptr,
+                 ParseInstructionForwarder, &diag);
+  spvDiagnosticDestroy(diag);
+}
+
+std::string FriendlyNameMapper::NameForId(uint32_t id) {
+  std::string result;
+  auto iter = name_for_id_.find(id);
+  if (iter == name_for_id_.end()) {
+    // It must have been an invalid module, so just return a trivial mapping.
+    // We don't care about uniqueness.
+    result = to_string(id);
```

Comment:
Oh, wait, keeping a local `result` allows the compiler to do RVO, in which `result` will be constructed directly in the return value's memory? Then my previous argument is wrong... Sorry, the modified version is in now.